Exiled.API/Features/Player.cs

```diff
@@ -128,6 +128,21 @@ public Player(GameObject gameObject)
         /// </summary>
         public static Dictionary<int, Player> IdsCache { get; } = new(20);
 
+        /// <summary>
+        /// Gets or sets a <see cref="Dictionary{TKey, TValue}"/> containing cached <see cref="RoleType"/> and their FF multiplier. This is for non-unique roles.
+        /// </summary>
+        public Dictionary<RoleType, int> FriendlyFireRules { get; set; } = new();
+
+        /// <summary>
+        /// Gets or sets a <see cref="Dictionary{TKey, TValue}"/> containing cached <see cref="string"/> and their  <see cref="Dictionary{TKey, TValue}"/> which is cached Role with FF multiplier. This is for unique custom roles.
+        /// </summary>
+        public Dictionary<string, Dictionary<RoleType, int>> UniqueFriendlyFireRules { get; set; } = new();
+
+        /// <summary>
+        /// Gets or sets a unique custom role that does not adbide to base game for this player. Used in conjunction with <see cref="UniqueFriendlyFireRules"/>.
+        /// </summary>
+        public string UniqueRole { get; set; } = string.Empty;
```

Comment:
Why would this feature not just be integrated directly into CustomRoles where it belongs?

Reply:
Because a lot of people rather make custom roles themselves, and while I agree that it should all be consolidated to reduce issues - we live in reality where that is not necessarily going to happen. Therefore, it is better to provide some level of accessibility rather than not doing anything at all and having 200x versions, at least if I give some level of control, it reduces how far the variants go, hopefully.

Reply:
This is 1. Stupid and 2. Counterintuitive
Firstly, you don't even incorporate this "UniqueRole" stuff in the existing CustomRoles API process. At the very least, player.UniqueRole should be set when they gain/lose a CustomRole.
Secondly, I don't really care what tinybrains decide to do for implementing their own custom roles, the EXILED method of doing it involves using the CustomRole API, and any code that is related to "custom player roles" should be contained soley within that API unless absolutely necessary for EXILED to function.

Reply:
1. Would it not be your job, the creator of CustomRoles to incorporate it? This system is independent of that system, and would be accessibly by those components. If it is my job since I adding a feature, I suppose. However, they are not at the moment connected.
2. I disagree that users are tinybrain for doing their own CustomRole API's - I do agree it is something that can create issues; however, the CustomRole code is not perfect, or may limit their intents and purposes. I cannot claim that to be fully true though, as I have not played with the CustomRole API itself.
3. While I understand your position, I do not believe this is Counterintuitive - it simply allows another facet for users to control their own player, and their FF rules. Even with CustomRoles, you need to be able to identify a unique role, which this supplies. Maybe I need to actually delve into the CustomRoles code, but it seems quiet efficient to provide this functionality whether or not it is directly folded into CustomRoles. 
4. With all I have previously said, I will not force this down anyone's throat. You tell me you do not want this, and it's gone. I will abide to the rules set whether they conflict with my own views or not. I simply am offering an alternative

Reply:
I have read through your CustomRole code, I understand your concerns a little bit more although there are a few arguments to justify why someone would not want to use this utility (A plugin, in essence). However, I see many benefits in this system itself.

Reply:
1. No because it is *you* adding the feature, so it is *you* who should incorporate it into the CustomRoles API that's included in EXILED.
2. Irrelevant. Since EXILED offers a CustomRole API, all of our 'CustomRole' related things should be contained within it.
3. They could manage this themselves via their plugin that they would need to make to add these rules anyways. Again, it seems like this is a feature that belongs in a plugin like Common Utilities, but is shoved into the API instead.
4. I'm not completely opposed to having the role-specific FF multipliers as part of the Player API, but I think the "custom role" stuff should be moved to and managed by the CustomRole API, as that would make much more sense in the context of a plugin properly using the EXILED APIs. If someone wants to make their own custom Role system, they can implement FF multipliers in that system themselves. It shoudn't be our responsibility to account for that, when we already provide a system for custom roles.  It would be akin to accounting for people making custom items without using our custom item API.

Reply:
1. Alright.. More work :( 
2. Sure
3. Alright Alright, I get it
4. Fair, I will fold it into your CustomRole code (basically just going to move the dict over, and then during transpiler check "Is this person a custom role, ok check their custom role dict". I will also copy over the helper functions into CustomRoles

Reply:
@joker-119 

My brain might be a little fried from fixing the last few things; however, let me know if this is dumb.

The minimum you stated was:
"Firstly, you don't even incorporate this "UniqueRole" stuff in the existing CustomRoles API process. At the very least, player.UniqueRole should be set when they gain/lose a CustomRole." 

and 

"I'm not completely opposed to having the role-specific FF multipliers as part of the Player API, but I think the "custom role" stuff should be moved to and managed by the CustomRole API, as that would make much more sense in the context of a plugin properly using the EXILED APIs. If someone wants to make their own custom Role system, they can implement FF multipliers in that system themselves. It shoudn't be our responsibility to account for that, when we already provide a system for custom roles. It would be akin to accounting for people making custom items without using our custom item API."

That means you will be at least alright with:

AddRole(){
....
// Something unique for the role since I assume there's no exact name for the custom role themselves
player.UniqueRole =  player.ReferenceHub.GetInstanceID().ToString();
}

Or something like that - If that is not accurate/dumb let me know. If that is the minimum you want, then everything else can remain the same, correct? If I were to actually move everything to CustomRoles then it would mean the transpiler would not exactly have access the information it needs. What I exactly mean is

Current code checks Player.(Custom..) and then does logic based off that. Because CustomRoles is a plugin, I cannot just do

Dict<CustomRoles, RoleRules..> in the Player class, meaning I would need to pass over the CustomRole object, at least to the player one, but I can't do that either because it is an external tool (and it should not be forced in).

Therefore, It makes sense that the role-specific FF stays a part of it, and CustomRoles just does initialization in a simpler way. Meaning, I would update custom roles to have the config file specify "What are the FF rules you will allow" similar to

  friendly_fire_rules:
    ClassD: 0
    Scp173: 1
    Scp93953: 2
    Scp93989: 3

So that for every custom role you define, you can specify in the config what rules should be associated. That's the only way I can think of right now on how to approach it. Let me know what you think, thanks